broker/src/main/java/io/zeebe/broker/logstreams/LogStreamDeletionService.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.0. You may not use this file
+ * except in compliance with the Zeebe Community License 1.0.
+ */
+package io.zeebe.broker.logstreams;
+
+import io.zeebe.broker.logstreams.state.StatePositionSupplier;
+import io.zeebe.logstreams.log.LogStream;
+import io.zeebe.logstreams.state.Snapshot;
+import io.zeebe.logstreams.state.SnapshotDeletionListener;
+import io.zeebe.logstreams.state.SnapshotStorage;
+import io.zeebe.servicecontainer.Injector;
+import io.zeebe.servicecontainer.Service;
+import io.zeebe.servicecontainer.ServiceStartContext;
+import io.zeebe.servicecontainer.ServiceStopContext;
+import io.zeebe.util.sched.Actor;
+import java.util.Objects;
+
+public class LogStreamDeletionService extends Actor
+    implements Service<SnapshotDeletionListener>, SnapshotDeletionListener {
+  private final Injector<SnapshotStorage> snapshotStorageInjector = new Injector<>();
+  private final LogStream logStream;
+  private final StatePositionSupplier positionSupplier;
+
+  private boolean isClosing;
+  private SnapshotStorage snapshotStorage;
+
+  public LogStreamDeletionService(
+      final LogStream logStream, final StatePositionSupplier positionSupplier) {
+    this.logStream = logStream;
+    this.positionSupplier = positionSupplier;
+  }
+
+  @Override
+  public void start(final ServiceStartContext startContext) {
+    snapshotStorage =
+        Objects.requireNonNull(
+            getSnapshotStorageInjector().getValue(), "must have a snapshot storage");
+    startContext.async(startContext.getScheduler().submitActor(this));
+  }
+
+  @Override
+  public void stop(final ServiceStopContext stopContext) {
+    stopContext.async(actor.close());
+  }
+
+  @Override
+  public SnapshotDeletionListener get() {
+    return this;
+  }
+
+  @Override
+  protected void onActorStarting() {
+    isClosing = false;
+    snapshotStorage.addDeletionListener(get());
+  }
+
+  @Override
+  protected void onActorClosing() {
+    isClosing = true;
+
+    if (snapshotStorage != null) {
+      snapshotStorage.removeDeletionListener(get());
+    }
+  }
+
+  @Override
+  public void onSnapshotDeleted(final Snapshot snapshot) {
```

Comment:
What snapshot is this? From reading the method definition it sounds like it is the deleted snapshot, which I assume is not the case.

Reply:
I think the doc mentions the argument is the oldest remaining snapshot, but maybe that's not clear enough. It's meant to be a callback as in "snapshots were deleted, here's the oldest remaining snapshot". If you have a better name idea I'll gladly change it :sweat_smile:

Reply:
hm like `afterSnapshotCleanUpVisitOldestRemainingSnapshot` :laughing: maybe we can rename the snapshot parameter as `oldesRemainingSnapshot`

Reply:
I went with `onSnapshotsDeleted(Snapshot oldestRemainingSnapshot)` (I think it's maybe clearer the arg is not the deleted snapshot if the method is plural)